pkg/storage/sqlstorage/balances.go

```diff
@@ -0,0 +1,210 @@
+	switch s.Schema().Flavor() {
+	case sqlbuilder.PostgreSQL:
+		sb.Select("account", "array_agg((asset, input - output))")
+	case sqlbuilder.SQLite:
+		// we try to get the same format as array_agg from postgres : {"(USD,-12686)","(EUR,-250)"}
+		// so don't have to dev a marshal method for each storage
+		sb.Select("account", `'{"(' || group_concat(asset||','||(input-output), ')","(')|| ')"}' as key_value_pairs`)
+	}
+
+	sb.From(s.schema.Table("volumes"))
+	sb.GroupBy("account")
+
+	t := BalancesPaginationToken{}
+
+	if q.AfterAddress != "" {
+		sb.Where(sb.L("account", q.AfterAddress))
+		t.AfterAddress = q.AfterAddress
+	}
+
+	if q.Filters.Address != "" {
+		arg := sb.Args.Add("^" + q.Filters.Address + "$")
+		switch s.Schema().Flavor() {
+		case sqlbuilder.PostgreSQL:
+			sb.Where("account ~* " + arg)
+		case sqlbuilder.SQLite:
+			sb.Where("account REGEXP " + arg)
+		}
+		t.AddressRegexpFilter = q.Filters.Address
+	}
+
+	sb.Limit(int(q.Limit + 1))
+	t.Limit = q.Limit
+	sb.Offset(int(q.Offset))
+
+	balanceQuery, args := sb.BuildWithFlavor(s.schema.Flavor())
+	rows, err := exec.QueryContext(ctx, balanceQuery, args...)
+	if err != nil {
+		return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+	}
+
+	defer func(rows *sql.Rows) {
+		err := rows.Close()
+		if err != nil {
+			panic(err)
+		}
+	}(rows)
+
+	accounts := make([]core.AccountsBalances, 0)
+
+	for rows.Next() {
+		var (
+			currentAccount string
+
+			assetBalance []byte
+		)
+		err = rows.Scan(&currentAccount, &assetBalance)
+		if err != nil {
+			return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+		}
+
+		// we (clean and) marshal the computed row in a map[string]int64
+		assetBalances := arrayToAssetBalance(assetBalance)
+		if err != nil {
+			return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+		}
+
+		accounts = append(accounts, map[string]map[string]int64{
+			currentAccount: assetBalances,
+		})
+	}
+
+	if err := rows.Err(); err != nil {
+		return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+	}
+
+	var previous, next string
+	if q.Offset > 0 {
+		offset := int(q.Offset) - int(q.Limit)
+		if offset < 0 {
+			t.Offset = 0
+		} else {
+			t.Offset = uint(offset)
+		}
+		raw, err := json.Marshal(t)
+		if err != nil {
+			return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+		}
+		previous = base64.RawURLEncoding.EncodeToString(raw)
+	}
+
+	if len(accounts) == int(q.Limit+1) {
+		accounts = accounts[:len(accounts)-1]
+		t.Offset = q.Offset + q.Limit
+		raw, err := json.Marshal(t)
+		if err != nil {
+			return sharedapi.Cursor[core.AccountsBalances]{}, s.error(err)
+		}
+		next = base64.RawURLEncoding.EncodeToString(raw)
+	}
+
+	return sharedapi.Cursor[core.AccountsBalances]{
+		PageSize: len(accounts),
+		Previous: previous,
+		Next:     next,
+		Data:     accounts,
+	}, nil
+}
+
+func (s *Store) GetAggregatedBalancesData(ctx context.Context, exec executor, q storage.BalancesQuery) (map[string]int64, error) {
```

Comment:
To improve code clarity:
```suggestion
func (s *Store) GetAggregatedBalancesData(ctx context.Context, exec executor, q storage.BalancesQuery) (core.AssetsBalances, error) {
```